functions/venue.js

```diff
@@ -476,23 +476,36 @@ exports.createVenue_v2 = functions.https.onCall(async (data, context) => {
 
 exports.upsertRoom = functions.https.onCall(async (data, context) => {
   checkAuth(context);
+
+  const result = {
+    provided: data,
+  };
+
   const { venueId, roomIndex, room } = data;
   await checkUserIsOwner(venueId, context.auth.token.user_id);
   const doc = await admin.firestore().collection("venues").doc(venueId).get();
 
   if (!doc || !doc.exists) {
     throw new HttpsError("not-found", `Venue ${venueId} not found`);
   }
-  const docData = doc.data();
-  let rooms = docData.rooms;
 
-  if (typeof roomIndex !== "number") {
-    rooms = [...rooms, room];
+  result.rooms = doc.data().rooms ?? [];
+
+  if (typeof roomIndex === "number") {
+    result.rooms[roomIndex] = room;
+    result.roomIndex = roomIndex;
   } else {
-    rooms[roomIndex] = room;
+    result.rooms = [...result.rooms, room];
+    result.roomIndex = result.rooms.length - 1;
   }
 
-  admin.firestore().collection("venues").doc(venueId).update({ rooms });
+  result.writeResult = await admin
+    .firestore()
+    .collection("venues")
+    .doc(venueId)
+    .update({ rooms: result.rooms });
+
+  return result;
```

Comment:
NOTE: None of the few places `upsertRoom` was used were handling any return value, so it's safe to add one now for future use.